skfp/distances/all_bit.py

```diff
@@ -0,0 +1,365 @@
+    >>> from skfp.distances import all_bit_count_similarity
+    >>> from scipy.sparse import csr_array
+    >>> vec_a = csr_array([[7, 1, 1]])
+    >>> vec_b = csr_array([[7, 1, 1]])
+    >>> sim = all_bit_count_similarity(vec_a, vec_b)
+    >>> sim  # doctest: +SKIP
+    1.0
+    """
+    _check_nan(vec_a)
+    _check_nan(vec_b)
+
+    if np.sum(vec_a) == 0 and np.sum(vec_b) == 0:
+        return 1.0
+
+    if isinstance(vec_a, csr_array) and isinstance(vec_b, csr_array):
+        return _all_bit_count_scipy(vec_a, vec_b)
+    elif isinstance(vec_a, np.ndarray) and isinstance(vec_b, np.ndarray):
+        return _all_bit_count_numpy(vec_a, vec_b)
+    else:
+        raise TypeError(
+            f"Both vec_a and vec_b must be of the same type, either numpy.ndarray "
+            f"or scipy.sparse.csr_array, got {type(vec_a)} and {type(vec_b)}"
+        )
+
+
+@validate_params(
+    {
+        "vec_a": ["array-like", csr_array],
+        "vec_b": ["array-like", csr_array],
+    },
+    prefer_skip_nested_validation=True,
+)
+def all_bit_count_distance(
+    vec_a: Union[np.ndarray, csr_array], vec_b: Union[np.ndarray, csr_array]
+) -> float:
+    """
+    All-Bit distance for vectors of count values.
+
+    Computes the All-Bit distance for count data between two input arrays
+    or sparse matrices by subtracting the similarity from 1, using to
+    the formula:
+
+    .. math::
+
+        dist(vec_a, vec_b) = 1 - sim(vec_a, vec_b)
+
+    The calculated distance falls within the range ``[0, 1]``.
+    Passing all-zero vectors to this function results in a distance of 0.
+
+    Parameters
+    ----------
+    vec_a : {ndarray, sparse matrix}
+        First count input array or sparse matrix.
+
+    vec_b : {ndarray, sparse matrix}
+        Second count input array or sparse matrix.
+
+    Returns
+    -------
+    distance : float
+        All-Bit distance between vec_a and vec_b.
+
+    Examples
+    --------
+    >>> from skfp.distances import all_bit_count_distance
+    >>> import numpy as np
+    >>> vec_a = np.array([1, 0, 1])
+    >>> vec_b = np.array([1, 0, 1])
+    >>> dist = all_bit_count_distance(vec_a, vec_b)
+    >>> dist  # doctest: +SKIP
+    0.0
+
+    >>> from skfp.distances import all_bit_count_distance
+    >>> from scipy.sparse import csr_array
+    >>> vec_a = csr_array([[1, 0, 1]])
+    >>> vec_b = csr_array([[1, 0, 1]])
+    >>> dist = all_bit_count_distance(vec_a, vec_b)
+    >>> dist  # doctest: +SKIP
+    0.0
+    """
+    return 1 - all_bit_count_similarity(vec_a, vec_b)
+
+
+@njit(parallel=True)
+def _all_bit_binary_numpy(vec_a: np.ndarray, vec_b: np.ndarray) -> float:
+    count_a = np.count_nonzero(vec_a)
+    xor_result = np.empty_like(vec_a, dtype=vec_a.dtype)
+
+    for i in prange(vec_a.size):
+        xor_result[i] = vec_a[i] ^ vec_b[i]
+
+    count_xor = np.count_nonzero(xor_result)
+
+    if count_a == 0:
+        return 0.0
+
+    all_bit_sim = (count_a - count_xor) / count_a
+    return max(0.0, all_bit_sim)
```

Comment:
This can be made much more efficiently for binary vectors, see SciPy code https://github.com/scipy/scipy/blob/v1.15.1/scipy/spatial/distance.py#L910, and also RDKit code https://github.com/rdkit/rdkit/blob/master/Code/DataStructs/BitOps.cpp#L543. You can use bitwise operations when casting to boolean vectors. This is basically `a AND b / len(a)`